File: architecture-skills/design-patterns/examples/factory-method/factory-method.c
/* Factory Method pattern in C.
 *
 * C has no classes or inheritance, so the creator is emulated as a struct
 * holding a `create` function pointer. The shared workflow function
 * (plan_delivery) calls through the pointer, and each concrete creator
 * supplies its own constructor.
 */

#include <stdio.h>

/* Product: a transport with its own deliver behavior. */
typedef struct Transport {
    const char *name;
    void (*deliver)(const struct Transport *self, const char *cargo);
} Transport;

static void truck_deliver(const Transport *self, const char *cargo) {
    printf("%s delivers %s by road in a box\n", self->name, cargo);
}

static void ship_deliver(const Transport *self, const char *cargo) {
    printf("%s delivers %s by sea in a container\n", self->name, cargo);
}

static Transport create_truck(void) {
    Transport t = { "Truck", truck_deliver };
    return t;
}

static Transport create_ship(void) {
    Transport t = { "Ship", ship_deliver };
    return t;
}

/* Creator: the factory method is a function pointer. */
typedef struct {
    const char *name;
    Transport (*create_transport)(void);
} Logistics;

static const Logistics ROAD_LOGISTICS = { "RoadLogistics", create_truck };
static const Logistics SEA_LOGISTICS = { "SeaLogistics", create_ship };

/* Shared workflow: uses whatever transport the creator produces. */
static void plan_delivery(const Logistics *logistics, const char *cargo) {
    Transport transport = logistics->create_transport();
    printf("[%s] ", logistics->name);
    transport.deliver(&transport, cargo);
}

int main(void) {
    plan_delivery(&ROAD_LOGISTICS, "machine parts");
    plan_delivery(&SEA_LOGISTICS, "machine parts");
    return 0;
}

File: architecture-skills/design-patterns/examples/factory-method/factory-method.rs
// Factory Method pattern in Rust.
//
// The creator is a trait whose default method (plan_delivery) is the shared
// workflow; implementors supply only create_transport. For a closed set of
// products, an enum with a constructor match is often preferred over
// boxed trait objects.

trait Transport {
    fn deliver(&self, cargo: &str) -> String;
}

struct Truck;
struct Ship;

impl Transport for Truck {
    fn deliver(&self, cargo: &str) -> String {
        format!("Truck delivers {cargo} by road in a box")
    }
}

impl Transport for Ship {
    fn deliver(&self, cargo: &str) -> String {
        format!("Ship delivers {cargo} by sea in a container")
    }
}

/// Creator: the default method is the shared workflow, the required
/// method is the factory step.
trait Logistics {
    fn name(&self) -> &'static str;
    fn create_transport(&self) -> Box<dyn Transport>;

    fn plan_delivery(&self, cargo: &str) -> String {
        let transport = self.create_transport();
        format!("[{}] {}", self.name(), transport.deliver(cargo))
    }
}

struct RoadLogistics;
struct SeaLogistics;

impl Logistics for RoadLogistics {
    fn name(&self) -> &'static str {
        "RoadLogistics"
    }
    fn create_transport(&self) -> Box<dyn Transport> {
        Box::new(Truck)
    }
}

impl Logistics for SeaLogistics {
    fn name(&self) -> &'static str {
        "SeaLogistics"
    }
    fn create_transport(&self) -> Box<dyn Transport> {
        Box::new(Ship)
    }
}

fn main() {
    let all: [&dyn Logistics; 2] = [&RoadLogistics, &SeaLogistics];
    for logistics in all {
        println!("{}", logistics.plan_delivery("machine parts"));
    }
}
